tools/shell/linenoise/linenoise.cpp

```diff
@@ -87,7 +87,17 @@ TabCompletion Linenoise::TabComplete() const {
 	buf[pos] = '\0';
 	completionCallback(buf, &lc);
 	buf[pos] = prev_char;
-	result.completions.reserve(lc.len);
+	if (lc.len == 0) {
+		// no completions found - return
+		return result;
+	}
+	result.completions.reserve(lc.len + 1);
+	// add the original text as the first completion
```

Comment:
This is definitely a huge improvement, but it does add one slight strange behavior:

If you now continue to press `<tab>` you eventually reach your original term again, but then when you need to press `<tab>` twice to start the cycle again (so the final tab has no effect).

Reply:
Thanks for the feedback - I've now made it so Tab cycles fully, and Shift+Tab just ends the completion if done on the first element